### ISSUE TITLE
resume training from previous epoch

### ISSUE THREAD

**Issue by Bhee** [fchollet/keras#1872]
I saved the model and weights after each epoch using callbacks.ModelCheckpoint. I want to train it again from the last epoch.
How to set the model.fit() command to start from the previous epoch?

**Comment by tboquet** [comment]
Do you want to do something special with the history? If not, you can just call `.fit` one or several times and you will be able to continue to train the model. If you want to continue the training in another process, you just have to load the weights and call `model.fit()`.

**Comment by Bhee** [comment]
when I call model.fit()  after loading models and weights , it showing epoch  = 1. If I stop the training at 100 epoch. I want to resume the training with epoch=101.

**Comment by ymcui** [comment]
I think it is no matter whether it SHOWs the training is at epoch = 1 or epoch = 101.
As far as I know, the model itself doesn't save the EPOCH information into model file.
If you have loaded the correct previous model (the model should have been saved with epoch number), it should be no problem on continuing your training process.

**Comment by Bhee** [comment]
thank u

**Comment by tboquet** [comment]
@ymcui is right, the label of `epoch` is only a name for the iterations in the current `fit`. Sorry when I said history I meant the history dictionary the `fit` method returns. I think #1868 is basically the same question. If you think it resolves your problem please close the issue!

**Comment by nithishdivakar** [comment]
But there is a problem with this approach. What about hyper parameters that change according to epoch, say learning rate with a decay. Just restarting it with fit method doesn't take that into account.

**Comment by lolongcovas** [comment]
yeah, this happens to me when I resume the training process by loading weights.

I was training resnet18 with imagenet dataset, the model saved the weights at 1st epoch with lr=0.1 at beginning. I stopped it, then tried the resume functionality, and it turns out that the model starts with the same lr=0.1, and the loss increase for each iteration. To set the lr to the state of the 1st epoch, I changed the lr according to SGD lr update func: lr = lr * (1. / (1+decay*iterations)), however, it didnt work, the loss sill increases, but slower than with lr=0.1. Probably I should still lower the lr, but I dont understand why the loss still increase even the lr is set accordantly.

**Comment by ywenlu** [comment]
Try the **initial_epoch** argument in **.fit** method.

**Comment by smhoang** [comment]
using initial_epoch didn't work in this case

**Comment by lewfish** [comment]
> But there is a problem with this approach. What about hyper parameters that change according to epoch, say learning rate with a decay. Just restarting it with fit method doesn't take that into account.

Setting the `initial_epoch` in `fit_generator` is not enough to solve this problem when using the `ReduceLROnPlateau` callback because there's no way for the callback to know what the learning rate should be without having the history of the previous (ie. before resuming training) epochs. Perhaps the callback constructor should have an optional `history` parameter that can be used to correctly initialize the learning rate and the `wait` variable (see https://github.com/fchollet/keras/blob/ab3b93e8dd103f1d9729305825791a084c7c8493/keras/callbacks.py#L744)

**Comment by MartinThoma** [comment]
Besides using the `initial_epoch` argument of `fit`, I re-wrote the history callback:

```
class History(Callback):
    """
    Callback that records events into a `History` object.

    This callback is automatically applied to
    every Keras model. The `History` object
    gets returned by the `fit` method of models.
    """

    def on_train_begin(self, logs=None):
        if not hasattr(self, 'epoch'):
            self.epoch = []
            self.history = {}

    def on_epoch_end(self, epoch, logs=None):
        logs = logs or {}
        self.epoch.append(epoch)
        for k, v in logs.items():
            self.history.setdefault(k, []).append(v)
```

This allows using the same callback and it just appends to the end. @fchollet should I post a pull request for this? It seems to me that this is more useful than the current behaviour of overwriting the
logs in `on_train_begin`.

**Comment by i3v** [comment]
@MartinThoma ,
One would probably need to replace [this line](https://github.com/fchollet/keras/blob/2.0.4/keras/engine/training.py#L1099) with
```
    if initial_epoch==0:
        self.history = cbks.History()
```
to make your suggestion work, right? I've tried to make this stuff work, and eventually ran into a feeling that [too many different things should be changed](https://github.com/fchollet/keras/issues/6697). What do you think?

**Comment by syedfaizalex** [comment]
if you want to resume from epoch 101 ,simply use "initial_epoch = 101" in model.fit().

initial_epoch: Epoch at which to start training (useful for resuming a previous training run)

**Comment by jperl** [comment]
Seems that tensorflow estimators also support resuming training. ["Since the state of the model is persisted (in model_dir=PATH above), the model will improve the more iterations you train it, until it settles"](https://developers.googleblog.com/2017/09/introducing-tensorflow-datasets.html)

**Comment by bupedroni** [comment]
Related question: What happens to all the gradient computations that rely on a history of the gradients (when momentum is present, such as in ADAM and most gradient descent algorithms)? Does the checkpoint store these as well? Thanks!

**Comment by valekar** [comment]
@bupedroni: As far as I know, every time I loaded the existing model, all the hyperparameters were set to default values. 

Best way to resume is to write a custom callback and store all the hyperparameters and then start the training as mentioned by @MartinThoma

**Comment by morenoh149** [comment]
@MartinThoma I'd like a pull request implementing that, basically I'm training a model, but if I notice that the metrics haven't diverged I'd like to train for another x epochs. And also be able to plot the history overall in an additive way.

**Comment by imranparuk** [comment]
Still have this issues... any update on it?

**Comment by imranparuk** [comment]
just port your code to pytorch :laughing:

**Comment by nithishdivakar** [comment]
Ya. That actually worked for me. 2 years and counting.

**Comment by rushic24** [comment]
> I think it is no matter whether it SHOWs the training is at epoch = 1 or epoch = 101.
> As far as I know, the model itself doesn't save the EPOCH information into model file.
> If you have loaded the correct previous model (the model should have been saved with epoch number), it should be no problem on continuing your training process.

So does that mean if i call
model.fit(epochs = 20)

and 

model.fit(epochs=5)
model.fit(epochs=5)
model.fit(epochs=5)
model.fit(epochs=5)

both are same ??

**Comment by srcolinas** [comment]
> > I think it is no matter whether it SHOWs the training is at epoch = 1 or epoch = 101.
> > As far as I know, the model itself doesn't save the EPOCH information into model file.
> > If you have loaded the correct previous model (the model should have been saved with epoch number), it should be no problem on continuing your training process.
> 
> So does that mean if i call
> model.fit(epochs = 20)
> 
> and
> 
> model.fit(epochs=5)
> model.fit(epochs=5)
> model.fit(epochs=5)
> model.fit(epochs=5)
> 
> both are same ??

Yes, they are equivalent. At least that is what I found using the TensorFlow Keras API in TensorFlow 2.0

**Comment by MunishaTripping** [comment]
How can I get the epoch at which model was saved in ModelCheckpoint ?

**Comment by hollowgalaxy** [comment]
save epoch number in the name of the model. Fetch that number with regex when resuming training.

**Comment by MichelHalmes** [comment]
I managed to do this with an optimizer whose learning rate depends on the number of iterations eg Adam.

Here is the pseudo-code:
```python
...
if os.path.isfile(checkpoint_path+".index"):
    # This loads `(root).optimizer.iter`from the checkpoint
    model.load_weights(checkpoint_path)

# Recover the iterations from the model and conver to epochs
initial_epoch = model.optimizer.iterations.numpy() // STEPS_PER_EPOCH
callback = tf.keras.callbacks.ModelCheckpoint(filepath=checkpoint_path, save_weights_only=True)
model.fit(train_data, epochs=NUM_EPOCHS, initial_epoch=initial_epoch,
                 callbacks=[callback])
```

Hope this helps :-)

**Comment by dorukkarinca** [comment]
I got tired of this so I ended up writing my own Keras wrapper that autoaaves and restores the epoch number, history, and model weights:

`pip install keras-buoy`
[Github project](https://github.com/dorukkarinca/keras-buoy)

Let me know what you think. PRs more than welcome.

**Comment by morenoh149** [comment]
@dorukkarinca is this handled in tensor flow v2? that's supposed to supersede keras

**Comment by dorukkarinca** [comment]
@morenoh149 not to the best of my knowledge. This wrapper wraps tensorflow.keras anyway.

**Comment by KazegamiKuon** [comment]
@dorukkarinca  UwU and Orz . Your wrapper help me so much. I didnt know why I wasted 1 week to retrain at start.

**Comment by RoyiAvital** [comment]
> So does that mean if i call
> model.fit(epochs = 20)
>
> and
> 
> model.fit(epochs=5)
> model.fit(epochs=5)
> model.fit(epochs=5)
> model.fit(epochs=5)

While they are the same, is there a simple way to append the history of each call in order to have the same history for both cases as well?

**Comment by Moshiur-Rahman-Faisal** [comment]
just use
callback = tf.keras.callbacks.experimental.BackupAndRestore(
backup_dir="temp")

**Comment by SExpert12** [comment]
Hello,
I am loading using function. Now how to access in fit to resume model training

def load_data(labels_file, test_size):
    """
    Display a list of images in a single figure with matplotlib.
        Parameters:
            labels_file: The labels CSV file.
            test_size: The size of the testing set.
    """
    labels = pd.read_csv(labels_file)
    X = labels[['center', 'left', 'right']].values
    y = labels['steering'].values
    X_train, X_valid, y_train, y_valid = train_test_split(X, y, test_size=test_size, random_state=0)
    return X_train, X_valid, y_train, y_valid

![image](https://user-images.githubusercontent.com/113903215/208099652-602bd156-ea09-4b14-9f68-cf703042a7c5.png)
![resumemodel_error](https://user-images.githubusercontent.com/113903215/208099775-1302093d-e38b-4044-82ea-292410885f24.png)

Please help me out.

**Comment by MoshiurRahmanFaisal** [comment]
Did you declared any X_train? Or divided your data set into train test? The
error shows in this screenshot tells me that you haven't declared any
X_train there your features for training data will be.

**Comment by MoshiurRahmanFaisal** [comment]
First fix your issue with the train and test and then declare a callback
function as bellow.

callback = tf.keras.callbacks.experimental.BackupAndRestore(
backup_dir="temp")

Now use this callback function while fitting your model. This will help you
to resume your training where you have left it.

Thank you and all the best.

**Comment by SExpert12** [comment]
Thanks for reply. Yes it is declared here.
![image](https://user-images.githubusercontent.com/113903215/208230375-cb410584-5cc9-4bb3-bd66-83941ba2c793.png)
Now how to get local variable value to access?

**Comment by MoshiurRahmanFaisal** [comment]
Use:
X_train, X_valid, y_train, y_valid = load_data(your labels file, your test
size)

Then try to fit your model on the same code where history = model.fit(your
code)

**Comment by SExpert12** [comment]
Okay. Let me try this out. Thanks for quick reply.

**Comment by SExpert12** [comment]
When I need to resume training I needed to start by execution of only this cell code:
![image](https://user-images.githubusercontent.com/113903215/208279210-e97d64e4-c1e6-4541-9543-fd97b3585025.png)

Am I right?

**Comment by SExpert12** [comment]
I am making some mistake in execution sequence or something. This is my code:
I execute these two cells first.
![image](https://user-images.githubusercontent.com/113903215/208279876-eff81b21-6d78-45c7-adf8-645731cbbb74.png)
then I execute this:
![image](https://user-images.githubusercontent.com/113903215/208279885-1acee0f0-da96-458b-8e31-8f63731e266a.png)
I stopped model training after some time to check it get resumed or not. After stopping model training, I execute this cell:
![image](https://user-images.githubusercontent.com/113903215/208279928-e7435361-c45f-44dc-b35d-0872dd83fbcf.png)
And I got error:
![image](https://user-images.githubusercontent.com/113903215/208279952-31bb28dc-5d54-488d-abee-8a932fe9bfb9.png)
![image](https://user-images.githubusercontent.com/113903215/208279962-83a067db-b6c1-4603-94c0-cadb7f235893.png)

Where am I making mistake?
